actions/auth-pubtkt

```diff
@@ -0,0 +1,138 @@
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Module with utilities to generate a auth_pubtkt ticket and
+sign it with the FreedomBox server's private key.
+"""
+
+import os
+import time
+import base64
+import datetime
+import argparse
+
+from OpenSSL import crypto
+
+KEYS_DIRECTORY = '/etc/apache2/auth-pubtkt-keys'
+
+
+def parse_arguments():
+    """ Return parsed command line arguments as dictionary. """
+    parser = argparse.ArgumentParser()
+    subparsers = parser.add_subparsers(dest='subcommand', help='Sub command')
+
+    subparsers.add_parser(
+        'create-key-pair',
+        help='create a key pair for the apache server '
+        'to sign auth_pubtkt tickets')
+    gen_tkt = subparsers.add_parser('generate-ticket',
+                                     help='generate auth_pubtkt ticket')
+    gen_tkt.add_argument('--uid', help='username of the user')
+    gen_tkt.add_argument('--private-key-file',
+                         help='path of the private key file of the server')
+    gen_tkt.add_argument('--tokens',
+                         help='tokens, usually containing the user groups')
+
+    subparsers.required = True
+    return parser.parse_args()
+
+
+def subcommand_create_key_pair(_):
+    """Create public/private key pair for signing the auth_pubtkt
+    tickets.
+    """
+    private_key_file = os.path.join(KEYS_DIRECTORY, 'privkey.pem')
+    public_key_file = os.path.join(KEYS_DIRECTORY, 'pubkey.pem')
+
+    os.path.exists(KEYS_DIRECTORY) or os.mkdir(KEYS_DIRECTORY)
+
+    if not all([
+            os.path.exists(key_file)
+            for key_file in [public_key_file, private_key_file]
+    ]):
+        pkey = crypto.PKey()
+        pkey.generate_key(crypto.TYPE_DSA, 1024)
+
+        with open(private_key_file, 'w') as priv_key_file:
+            priv_key = crypto.dump_privatekey(crypto.FILETYPE_PEM,
+                                              pkey).decode()
+            priv_key_file.write(priv_key)
+
+        with open(public_key_file, 'w') as pub_key_file:
+            pub_key = crypto.dump_publickey(crypto.FILETYPE_PEM, pkey).decode()
+            pub_key_file.write(pub_key)
+
+        for fil in [public_key_file, private_key_file]:
+            os.chmod(fil, 0o440)
+
+
+def create_ticket(pkey, uid, validuntil, ip=None, tokens=None,
+                  udata=None, graceperiod=None, extra_fields=None):
+    """Create and return a signed mod_auth_pubtkt ticket."""
+    fields = [
+        'uid={}'.format(uid),
+        'validuntil={}'.format(validuntil, type='d'),
+        ip and 'cip={}'.format(ip),
+        tokens and 'tokens={}'.format(tokens),
+        graceperiod and 'graceperiod={}'.format(graceperiod, type='d'),
+        udata and 'udata={}'.format(udata),
+        extra_fields and ';'.join(
+            ['{k}={v}'.format(k, v) for k, v in extra_fields])
```

Comment:
The named format fields {k} and {v} are not specified. Should just be '{}={}' to use positional fields.